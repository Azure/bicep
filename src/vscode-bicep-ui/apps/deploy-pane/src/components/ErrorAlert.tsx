// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import type { FC } from "react";

import { Codicon } from "@vscode-bicep-ui/components";
import styled from "styled-components";

const ErrorAlertDiv = styled.div`
  color: var(--vscode-statusBarItem-errorForeground);
  background-color: var(--vscode-statusBarItem-errorBackground);
  padding: 5px 10px;
  border-radius: 4px;
  font-size: 14px;
  align-self: center;
`;

export const ErrorAlert: FC<{ message: string }> = ({ message }) => {
  return (
    <ErrorAlertDiv>
      <Codicon name="error" size={14} />
      {message}
    </ErrorAlertDiv>
  );
};
