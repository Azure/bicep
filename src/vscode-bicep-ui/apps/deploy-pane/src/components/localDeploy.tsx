// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import type { FC } from "react";
import type { LocalDeploymentOperationContent, LocalDeployResponse } from "../messages";

import {
  VscodeTable,
  VscodeTableBody,
  VscodeTableCell,
  VscodeTableHeader,
  VscodeTableHeaderCell,
  VscodeTableRow,
} from "@vscode-elements/react-elements";
import { FormSection } from "./sections/FormSection";
import { getPreformattedJson } from "./utils";

export const LocalDeployResult: FC<{ result: LocalDeployResponse }> = ({ result }) => {
  const error = result.deployment.error;
  return (
    <FormSection title="Result">
      <p>{result.deployment.provisioningState}</p>
      {error && (
        <VscodeTable>
          {error && (
            <VscodeTableRow key={0}>
              <VscodeTableCell id="1">Code</VscodeTableCell>
              <VscodeTableCell id="2">{error.code}</VscodeTableCell>
            </VscodeTableRow>
          )}
          {error.message && (
            <VscodeTableRow key={1}>
              <VscodeTableCell id="1">Message</VscodeTableCell>
              <VscodeTableCell id="2">{error.message}</VscodeTableCell>
            </VscodeTableRow>
          )}
        </VscodeTable>
      )}
    </FormSection>
  );
};

export const LocalDeployOperations: FC<{ result: LocalDeployResponse }> = ({ result }) => {
  if (!result.operations) {
    return null;
  }

  return (
    <FormSection title="Operations">
      <VscodeTable>
        <VscodeTableHeader slot="header">
          <VscodeTableHeaderCell id="1">Resource Name</VscodeTableHeaderCell>
          <VscodeTableHeaderCell id="2">State</VscodeTableHeaderCell>
          <VscodeTableHeaderCell id="3">Error</VscodeTableHeaderCell>
        </VscodeTableHeader>
        <VscodeTableBody slot="body">
          {result.operations.map((operation) => (
            <VscodeTableRow
              key={operation.resourceName}
              style={isFailed(operation) ? { background: "rgba(255, 72, 45, 0.3)" } : {}}
            >
              <VscodeTableCell id="1">{operation.resourceName}</VscodeTableCell>
              <VscodeTableCell id="2">{operation.provisioningState}</VscodeTableCell>
              <VscodeTableCell id="3">{operation.error ? getPreformattedJson(operation.error) : ""}</VscodeTableCell>
            </VscodeTableRow>
          ))}
        </VscodeTableBody>
      </VscodeTable>
    </FormSection>
  );
};

export const LocalDeployOutputs: FC<{ result: LocalDeployResponse }> = ({ result }) => {
  if (!result.deployment.outputs) {
    return null;
  }

  return (
    <FormSection title="Outputs">
      <VscodeTable>
        <VscodeTableHeader slot="header">
          <VscodeTableHeaderCell id="1">Name</VscodeTableHeaderCell>
          <VscodeTableHeaderCell id="2">Value</VscodeTableHeaderCell>
        </VscodeTableHeader>
        <VscodeTableBody slot="body">
          {Object.keys(result.deployment.outputs).map((name) => (
            <VscodeTableRow key={name}>
              <VscodeTableCell id="1">{name}</VscodeTableCell>
              <VscodeTableCell id="2">{getPreformattedJson(result.deployment.outputs[name])}</VscodeTableCell>
            </VscodeTableRow>
          ))}
        </VscodeTableBody>
      </VscodeTable>
    </FormSection>
  );
};

function isFailed(operation: LocalDeploymentOperationContent) {
  return operation.provisioningState.toLowerCase() === "failed";
}
