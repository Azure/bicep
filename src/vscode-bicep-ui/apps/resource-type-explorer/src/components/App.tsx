// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { Accordion } from "@vscode-bicep-ui/components";
import { useWebviewRequest } from "@vscode-bicep-ui/messaging";
import { ResourceTypeGroupHeader } from "./ResourceTypeGroupHeader";
import { ResourceTypeGroupList } from "./ResourceTypeGroupList";

export type ResourceTypes = {
  resourceType: string;
  apiVersion: string;
};

type ResourceTypeCatalog = Array<{
  group: string;
  resourceTypes: ResourceTypes[];
}>;

export function App() {
  const [resourceTypeCatalog, error] = useWebviewRequest<ResourceTypeCatalog>("resourceTypeCatalog/load");

  if (error instanceof Error) {
    console.error(error);
  }

  return (
    <>
      {error && <div>Failed to load resource types</div>}
      {resourceTypeCatalog && (
        <section>
          <Accordion>
            {resourceTypeCatalog.map(({ group, resourceTypes }) => (
              <Accordion.Item key={group}>
                <Accordion.ItemCollapse>
                  <ResourceTypeGroupHeader group={group} />
                </Accordion.ItemCollapse>
                <Accordion.ItemContent>
                  <ResourceTypeGroupList group={group} resourceTypes={resourceTypes} />
                </Accordion.ItemContent>
              </Accordion.Item>
            ))}
          </Accordion>
        </section>
      )}
    </>
  );
}
