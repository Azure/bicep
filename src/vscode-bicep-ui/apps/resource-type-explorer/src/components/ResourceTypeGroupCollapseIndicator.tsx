// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { Codicon, useAccordionItem } from "@vscode-bicep-ui/components";

export function ResourceTypeGroupCollapseIndicator() {
  const { active } = useAccordionItem();

  return <Codicon name={active ? "chevron-down" : "chevron-right"} size={16} />;
}
