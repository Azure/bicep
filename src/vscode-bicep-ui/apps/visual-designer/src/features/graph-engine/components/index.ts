// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

export * from "./Canvas";
export * from "./Graph";
