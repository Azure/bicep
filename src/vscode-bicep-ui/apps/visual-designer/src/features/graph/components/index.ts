export * from "./Canvas";
export * from "./Graph";
