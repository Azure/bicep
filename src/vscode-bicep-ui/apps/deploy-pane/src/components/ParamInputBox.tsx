// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import type { FC, PropsWithChildren } from "react";
import type { ParamDefinition, ParameterValue } from "../models";

import {
  VscodeCheckbox,
  VscodeLabel,
  VscodeOption,
  VscodeSingleSelect,
  VscodeTextarea,
  VscodeTextfield,
} from "@vscode-elements/react-elements";
import { useState } from "react";
import styled from "styled-components";
import { ErrorAlert } from "./ErrorAlert";

export type ParameterInputData = {
  inputValue?: unknown;
  value?: ParameterValue;
  isValid: boolean;
};

type InputProps = {
  definition: ParamDefinition;
  disabled: boolean;
  data: ParameterInputData;
  onChangeData: (name: string, value: ParameterInputData) => void;
};

const getInputHtmlId = (definition: ParamDefinition) => `param-input-${definition.name.toLowerCase()}`;

const ParamCheckboxInput: FC<InputProps> = (props) => {
  const { definition, data, disabled, onChangeData } = props;
  const { name, defaultValue } = definition;
  const defaultInputValue = defaultValue as boolean;

  const [inputValue, setInputValue] = useState<boolean>((data.inputValue as boolean) ?? defaultInputValue);

  function handleChangeValue(value: boolean) {
    onChangeData(definition.name, { inputValue: value, value: value, isValid: true });
    setInputValue(value);
  }

  return (
    <InputBoxWrapper disabled={disabled}>
      <VscodeCheckbox
        id={getInputHtmlId(definition)}
        checked={inputValue}
        onChange={() => handleChangeValue(!inputValue)}
        disabled={disabled}
      >
        {name}
      </VscodeCheckbox>
    </InputBoxWrapper>
  );
};

const ParamIntInput: FC<InputProps> = (props) => {
  const { definition, data, disabled, onChangeData } = props;
  const { name, defaultValue } = definition;
  const defaultInputValue = `${defaultValue ?? 0}`;

  const [inputValue, setInputValue] = useState<string>((data.inputValue as string) ?? defaultInputValue);
  const [error, setError] = useState<string>();

  function handleChangeValue(value: string) {
    const newValueInt = Number(value);
    if (Number.isInteger(newValueInt)) {
      onChangeData(definition.name, { inputValue: value, value: newValueInt, isValid: true });
      setError(undefined);
    } else {
      onChangeData(definition.name, { inputValue: value, isValid: false });
      setError("Invalid integer value");
    }
    setInputValue(value);
  }

  return (
    <InputBoxWrapper disabled={disabled} error={error}>
      <VscodeLabel htmlFor={getInputHtmlId(definition)}>{name}</VscodeLabel>
      <VscodeTextfield
        id={getInputHtmlId(definition)}
        value={inputValue}
        onChange={(e) => handleChangeValue((e.currentTarget as HTMLInputElement).value)}
        disabled={disabled}
      />
    </InputBoxWrapper>
  );
};

const ParamStringInput: FC<InputProps> = (props) => {
  const { definition, data, disabled, onChangeData } = props;
  const { name, defaultValue } = definition;
  const defaultInputValue = defaultValue as string;

  const [inputValue, setInputValue] = useState<string>((data.inputValue as string) ?? defaultInputValue);

  function handleChangeValue(value: string) {
    setInputValue(value);
    onChangeData(definition.name, { inputValue: value, value: value, isValid: true });
  }

  if (definition.allowedValues) {
    return (
      <InputBoxWrapper disabled={disabled}>
        <VscodeLabel htmlFor={getInputHtmlId(definition)}>{name}</VscodeLabel>
        <VscodeSingleSelect
          id={getInputHtmlId(definition)}
          onChange={(e) => handleChangeValue((e.currentTarget as HTMLSelectElement).value)}
          disabled={disabled}
        >
          {definition.allowedValues.map((option) => (
            <VscodeOption key={option} selected={inputValue === option}>
              {option}
            </VscodeOption>
          ))}
        </VscodeSingleSelect>
      </InputBoxWrapper>
    );
  } else {
    return (
      <InputBoxWrapper disabled={disabled}>
        <VscodeLabel htmlFor={getInputHtmlId(definition)}>{name}</VscodeLabel>
        <VscodeTextfield
          id={getInputHtmlId(definition)}
          value={inputValue}
          onChange={(e) => handleChangeValue((e.currentTarget as HTMLInputElement).value)}
          disabled={disabled}
        />
      </InputBoxWrapper>
    );
  }
};

const ParamJsonInput: FC<InputProps> = (props) => {
  const { definition, data, disabled, onChangeData } = props;
  const { name, defaultValue } = definition;
  const defaultInputValue = defaultValue ? JSON.stringify(defaultValue, null, 2) : "";

  const [inputValue, setInputValue] = useState<string>((data.inputValue as string) ?? defaultInputValue);
  const [error, setError] = useState<string>();

  function handleChangeValue(value: string) {
    try {
      onChangeData(definition.name, {
        inputValue: value,
        value: value !== "" ? JSON.parse(value) : undefined,
        isValid: true,
      });
      setError(undefined);
    } catch {
      onChangeData(definition.name, { inputValue: value, isValid: false });
      setError("Invalid JSON value");
    }
    setInputValue(value);
  }

  return (
    <InputBoxWrapper disabled={disabled} error={error}>
      <VscodeLabel htmlFor={getInputHtmlId(definition)}>{name}</VscodeLabel>
      <VscodeTextarea
        id={getInputHtmlId(definition)}
        className="code-textarea-container"
        resize="vertical"
        value={inputValue}
        onChange={(e) => handleChangeValue((e.currentTarget as HTMLInputElement).value)}
        disabled={disabled}
      />
    </InputBoxWrapper>
  );
};

type InputBoxWrapperProps = PropsWithChildren<{
  disabled: boolean;
  error?: string;
}>;

const InputRowDiv = styled.div`
  display: grid;
  grid-auto-flow: row;
`;

const InputBoxWrapper: FC<InputBoxWrapperProps> = (props) => {
  const { error, children } = props;
  return (
    <InputRowDiv>
      {children}
      {error && <ErrorAlert message={error} />}
    </InputRowDiv>
  );
};

export const ParamInputBox: FC<InputProps> = (props) => {
  const { definition } = props;
  const { type } = definition;

  switch (type) {
    case "bool":
      return <ParamCheckboxInput {...props} />;
    case "int":
      return <ParamIntInput {...props} />;
    case "string":
      return <ParamStringInput {...props} />;
    default:
      return <ParamJsonInput {...props} />;
  }
};
