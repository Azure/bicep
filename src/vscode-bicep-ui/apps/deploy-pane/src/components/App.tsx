// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import type { FC } from "react";
import type { ParametersInputData } from "./sections/ParametersInputView";

import { VscodeButton, VscodeProgressRing } from "@vscode-elements/react-elements";
import { useState } from "react";
import { ErrorAlert } from "./ErrorAlert";
import { useAzure } from "./hooks/useAzure";
import { useMessageHandler } from "./hooks/useMessageHandler";
import { LocalDeployOperations, LocalDeployOutputs, LocalDeployResult } from "./localDeploy";
import { DeploymentOperationsView } from "./sections/DeploymentOperationsView";
import { DeploymentOutputsView } from "./sections/DeploymentOutputsView";
import { DeploymentScopeInputView } from "./sections/DeploymentScopeInputView";
import { FormSection } from "./sections/FormSection";
import { ParametersInputView } from "./sections/ParametersInputView";
import { ResultsView } from "./sections/ResultsView";
import { WhatIfChangesView } from "./sections/WhatIfChangesView";

export const App: FC = () => {
  const [errorMessage, setErrorMessage] = useState<string>();
  const [localDeployRunning, setLocalDeployRunning] = useState(false);
  const [parameters, setParameters] = useState<ParametersInputData>({});
  const messages = useMessageHandler({ setErrorMessage, setLocalDeployRunning });
  const azure = useAzure({
    scope: messages.scope,
    acquireAccessToken: messages.acquireAccessToken,
    templateMetadata: messages.templateMetadata,
    setErrorMessage,
  });
  const isRunning = azure.deployState.status === "running" || localDeployRunning;

  function handleEnableParamEditing() {
    messages.setParamsMetadata({ ...messages.paramsMetadata, sourceFilePath: undefined });
  }

  const parametersValid = Object.values(parameters).every((param) => param.isValid);
  const azureDisabled = !messages.scope || !messages.templateMetadata || isRunning || !parametersValid;

  async function handleDeployClick() {
    messages.publishTelemetry("deployPane/deploy", {});
    await azure.deploy(parameters);
  }

  async function handleValidateClick() {
    messages.publishTelemetry("deployPane/validate", {});
    await azure.validate(parameters);
  }

  async function handleWhatIfClick() {
    messages.publishTelemetry("deployPane/whatIf", {});
    await azure.whatIf(parameters);
  }

  async function handleLocalDeployClick() {
    messages.publishTelemetry("deployPane/localDeploy", {});
    messages.startLocalDeploy();
  }

  if (!messages.messageState.initialized) {
    return <VscodeProgressRing />;
  }

  const showLocalDeployControls =
    messages.messageState.localDeployEnabled &&
    // if there's an error, this'll cause sourceFilePath to be empty - we still want to show the controls to display the error
    (errorMessage || messages.paramsMetadata.sourceFilePath?.endsWith(".bicepparam"));

  return (
    <main id="webview-body">
      {!messages.messageState.localDeployEnabled && (
        <>
          <DeploymentScopeInputView scope={messages.scope} onPickScope={messages.pickScope} />

          <ParametersInputView
            parameters={messages.paramsMetadata}
            template={messages.templateMetadata}
            disabled={isRunning}
            onParametersChange={setParameters}
            onEnableEditing={handleEnableParamEditing}
            onPickParametersFile={messages.pickParamsFile}
          />

          <FormSection title="Actions">
            {errorMessage && <ErrorAlert message={errorMessage} />}
            <div className="controls">
              <VscodeButton onClick={handleDeployClick} disabled={azureDisabled}>
                Deploy
              </VscodeButton>
              <VscodeButton onClick={handleValidateClick} disabled={azureDisabled}>
                Validate
              </VscodeButton>
              <VscodeButton onClick={handleWhatIfClick} disabled={azureDisabled}>
                What-If
              </VscodeButton>
            </div>
          </FormSection>

          {messages.scope && (
            <>
              <ResultsView scope={messages.scope} deployState={azure.deployState} />
              <DeploymentOperationsView scope={messages.scope} operations={azure.operations} />
              <DeploymentOutputsView outputs={azure.outputs} />
              <WhatIfChangesView changes={azure.whatIfChanges} />
            </>
          )}
        </>
      )}

      {messages.messageState.localDeployEnabled && (
        <>
          {showLocalDeployControls && (
            <>
              <ParametersInputView
                parameters={messages.paramsMetadata}
                template={messages.templateMetadata}
                disabled={localDeployRunning}
                onParametersChange={setParameters}
                onEnableEditing={handleEnableParamEditing}
                onPickParametersFile={messages.pickParamsFile}
              />

              <FormSection title="Actions">
                {errorMessage && <ErrorAlert message={errorMessage} />}
                <div className="controls">
                  <VscodeButton onClick={handleLocalDeployClick} disabled={localDeployRunning}>
                    Deploy
                  </VscodeButton>
                </div>
                {localDeployRunning && <VscodeProgressRing></VscodeProgressRing>}
              </FormSection>

              {!localDeployRunning && messages.localDeployResult && (
                <>
                  <LocalDeployResult result={messages.localDeployResult} />
                  <LocalDeployOperations result={messages.localDeployResult} />
                  <LocalDeployOutputs result={messages.localDeployResult} />
                </>
              )}
            </>
          )}
          {!showLocalDeployControls && (
            <ErrorAlert message="Local Deployment is only currently supported for .bicepparam files. Relaunch this pane for a .bicepparam file." />
          )}
        </>
      )}
    </main>
  );
};
