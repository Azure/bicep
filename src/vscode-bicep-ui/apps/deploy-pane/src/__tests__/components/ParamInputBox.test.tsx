// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import type { ParameterInputData } from "../../components/ParamInputBox";
import type { ParamDefinition } from "../../models";

import { render } from "@testing-library/react";
import { expect, it, vi } from "vitest";
import { ParamInputBox } from "../../components/ParamInputBox";

// TODO these tests aren't particularly meaningful, as vite doesn't support web components

it("renders a string input", () => {
  const definition: ParamDefinition = {
    name: "myStringParam",
    type: "string",
    defaultValue: "fooValue",
  };
  const data: ParameterInputData = {
    isValid: true,
  };
  const onChangeData = vi.fn();

  const { container } = render(
    <ParamInputBox definition={definition} data={data} disabled={false} onChangeData={onChangeData} />,
  );

  expect(container).toMatchSnapshot();
});

it("renders an integer input", () => {
  const definition: ParamDefinition = {
    name: "myIntParam",
    type: "int",
    defaultValue: 42,
  };
  const data: ParameterInputData = {
    isValid: true,
  };
  const onChangeData = vi.fn();

  const { container } = render(
    <ParamInputBox definition={definition} data={data} disabled={false} onChangeData={onChangeData} />,
  );

  expect(container).toMatchSnapshot();
});

it("renders a bool input", () => {
  const definition: ParamDefinition = {
    name: "myBoolParam",
    type: "bool",
    defaultValue: true,
  };
  const data: ParameterInputData = {
    isValid: true,
  };
  const onChangeData = vi.fn();

  const { container } = render(
    <ParamInputBox definition={definition} data={data} disabled={false} onChangeData={onChangeData} />,
  );

  expect(container).toMatchSnapshot();
});

it("renders a JSON input", () => {
  const definition: ParamDefinition = {
    name: "myJsonParam",
    type: "object",
    defaultValue: { foo: "bar" },
  };
  const data: ParameterInputData = {
    isValid: true,
  };
  const onChangeData = vi.fn();

  const { container } = render(
    <ParamInputBox definition={definition} data={data} disabled={false} onChangeData={onChangeData} />,
  );

  expect(container).toMatchSnapshot();
});
