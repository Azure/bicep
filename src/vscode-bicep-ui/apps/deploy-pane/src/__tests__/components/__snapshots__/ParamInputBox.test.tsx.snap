// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders a JSON input 1`] = `
<div>
  <div
    class="sc-kLhKbu iCmsqj"
  >
    <vscode-label
      id="vscode-label-3"
    >
      myJsonParam
    </vscode-label>
    <vscode-textarea
      class="code-textarea-container"
    />
  </div>
</div>
`;

exports[`renders a bool input 1`] = `
<div>
  <div
    class="sc-kLhKbu iCmsqj"
  >
    <vscode-checkbox>
      myBoolParam
    </vscode-checkbox>
  </div>
</div>
`;

exports[`renders a string input 1`] = `
<div>
  <div
    class="sc-kLhKbu iCmsqj"
  >
    <vscode-label
      id="vscode-label-1"
    >
      myStringParam
    </vscode-label>
    <vscode-textfield />
  </div>
</div>
`;

exports[`renders an integer input 1`] = `
<div>
  <div
    class="sc-kLhKbu iCmsqj"
  >
    <vscode-label
      id="vscode-label-2"
    >
      myIntParam
    </vscode-label>
    <vscode-textfield />
  </div>
</div>
`;
